### PR TITLE
powerdns-admin: fix and add module

### PR DESCRIPTION
**Pull request by Flakebi**
###### Motivation for this change
- Add the migrations directory to the package
- Add postgres support to the package
- Add a service for powerdns-admin

I use the service in combination with postgresql:
```nix
  services.powerdns-admin = {
    enable = true;
    config = ''
      import os
      SALT = os.environ["SALT"]
      SECRET_KEY = os.environ["SECRET_KEY"]

      BIND_ADDRESS = '127.0.0.1'
      PORT = 9000
      SQLALCHEMY_DATABASE_URI = 'postgresql://powerdnsadmin@/powerdnsadmin?host=/run/postgresql'
    '';
    extraArgs = [ "-b" "127.0.0.1:9000" ];
  };
  systemd.services.powerdns-admin = {
    after = [ "postgresql.service" ];
    serviceConfig.BindPaths = "/run/postgresql";
    serviceConfig.EnvironmentFile = "/etc/powerdnsadmin/secrets";
  };

```

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/powerdns-admin/default.nix [hunk 1]
```diff
@@ -91,6 +91,7 @@ in stdenv.mkDerivation rec {
 
   postPatch = ''
     rm -r powerdnsadmin/static powerdnsadmin/assets.py
+    sed -i "s/id:/'id':/" migrations/versions/787bdba9e147_init_db.py
```
Comment (SuperSandro2000) [inline pkgs/applications/networking/powerdns-admin/default.nix:94]: Why is this required?

Comment (Flakebi) [reply pkgs/applications/networking/powerdns-admin/default.nix:94]: Without this patch, running the migration fails with a `TypeError: '<' not supported between instances of 'str' and 'builtin_function_or_method'` because `id` is a function instead of a string.

**Comment by Flakebi** [comment]
Added a nixos test.

**Comment by zhaofengli** [comment]
Looks like the `SystemCallFilter` may be too strict. I'm using a MySQL database, and upon starting powerdns-admin immediately dumps core. The missing syscall is chown:

```
Dec 01 06:04:45 hostname audit[2965227]: SECCOMP auid=4294967295 uid=993 gid=990 ses=4294967295 subj=kernel pid=2965227 comm="python" exe="/nix/store/fkzla307l4mlcvfyshsrccwl7szihx2z-python3-3.9.6/bin/python3.9" sig=31 arch=c000003e syscall=92 compat=0 ip=0x7f56e9d1b407 code=0x80000000
```

But otherwise, great job! Thank you for doing this :heart:

File: nixos/modules/services/web-apps/powerdns-admin.nix [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        ProtectKernelModules = true;
+        ProtectKernelTunables = true;
+        ProtectProc = "invisible";
+        RestrictAddressFamilies = [ "AF_INET" "AF_INET6" "AF_UNIX" ];
+        RestrictNamespaces = true;
+        RestrictRealtime = true;
+        RestrictSUIDSGID = true;
+        SystemCallArchitectures = "native";
+        # gunicorn needs setuid
+        SystemCallFilter = [ "@system-service" "~@privileged @resources @keyring" "@setuid" ];
```
Comment (mweinelt) [inline nixos/modules/services/web-apps/powerdns-admin.nix:93]:
```suggestion
        SystemCallFilter = [ 
          "@system-service @setuid"
          "~@privileged @resources @keyring"
         ];
```

One allow and one deny list. Easier to parse that way.

Comment (Flakebi) [reply nixos/modules/services/web-apps/powerdns-admin.nix:93]: I just tried again, but this doesn’t work. The order is important.
1. Add `@system-service`
2. Remove `@privileged` and others, which includes `@setuid`
3. Readd `@setuid`

File: nixos/modules/module-list.nix [hunk 3]
```diff
@@ -1019,6 +1019,7 @@
   ./services/web-apps/plantuml-server.nix
   ./services/web-apps/plausible.nix
   ./services/web-apps/pgpkeyserver-lite.nix
+  ./services/web-apps/powerdns-admin.nix
```
Comment (zhaofengli) [inline nixos/modules/module-list.nix:1022]: We should also add a line to the release notes for the new module.

**Comment by Flakebi** [comment]
Thanks for testing @zhaofengli!
I added a test for mysql, but that seems to run through fine without chown permissions.

Good point of adding a changelog entry.

**Comment by zhaofengli** [comment]
> I added a test for mysql, but that seems to run through fine without chown permissions.

Aha, my setup required `@chown` because I was making it listen on a Unix socket (`extraArgs = [ "-b" "unix:%t/http.sock" ];`). I added tests for that case [here](https://github.com/zhaofengli/nixpkgs/commit/91858d2985664bad7d374c09d468257ab3cb307f). Could you cherry-pick? If you are testing on that commit, remove `@chown` from the syscall filter to make the test fail.

**Comment by zhaofengli** [comment]
Oh right, there is another thing which would be nice to add. Could you add an option `secretKeyFile` which requires the user to either specify a path to a file that contains the secret key (and inject a line to read the file in the config), or `null` if they intend to specify `SECRET_KEY` manually themselves.

This is important because if a `SECRET_KEY` is not specified, powerdns-admin will still start but with [the default key](https://github.com/ngoduykhanh/PowerDNS-Admin/blob/701a442d126ebd816d2a9728dba5d3a306ba0efb/powerdnsadmin/default_config.py) which allows anyone to trivially spoof a session. The README didn't even mention this [until someone sent a PR](ngoduykhanh/PowerDNS-Admin#931) earlier this year. We need to make this explicit so users can't accidentally spin up an insecure instance.

**Comment by Flakebi** [comment]
Nice find!
I cherry-picked your commit and added a `secretKeyFile` and `saltFile` option as you suggested.

File: nixos/modules/services/web-apps/powerdns-admin.nix [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+with lib;
+
+let
+  cfg = config.services.powerdns-admin;
+
+  configText = ''
+    ${cfg.config}
+  ''
+  + optionalString (cfg.secretKeyFile != null) ''
+    with open('${cfg.secretKeyFile}') as file:
```
Comment (zhaofengli) [inline nixos/modules/services/web-apps/powerdns-admin.nix:12]:
```suggestion
    with open('${cfg.secretKeyFile}', 'r') as file:
```

The service cannot and should not be able to write to the file (same for `saltFile`).

File: nixos/modules/services/web-apps/powerdns-admin.nix [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        Group = "powerdnsadmin";
+
+        AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        BindReadOnlyPaths = [
+          "/nix/store"
+          "-/etc/resolv.conf"
+          "-/etc/nsswitch.conf"
+          "-/etc/hosts"
+          "-/etc/localtime"
+        ];
```
Comment (zhaofengli) [inline nixos/modules/services/web-apps/powerdns-admin.nix:96]: We should add `secretKeyFile` and `saltFile` to `BindReadOnlyPaths` if they are specified, otherwise the service cannot read them from the isolated mount namespace.

**Review by zhaofengli — CHANGES_REQUESTED** [review]
Thanks! I just re-tested, and there are just a couple of final nits:

File: nixos/modules/services/web-apps/powerdns-admin.nix (hunk 5 above)
Comment (Flakebi) [reply nixos/modules/services/web-apps/powerdns-admin.nix]: Good point, I changed the tests to also check that.

**Review by zhaofengli — APPROVED** [review]
Nice! Thanks for all the work :heart:

**Comment by Flakebi** [comment]
Rebased to fix merge conflicts in the release notes.

**Review by zhaofengli — APPROVED** [review]
Can someone merge before more merge conflicts occur? There is another PR that I was planning to send to bump powerdns-admin depending on this PR merged as well.

Cc @SuperSandro2000 @mweinelt

**Comment by Flakebi** [comment]
Rebased again to fix merge conflicts in the release notes.

**Comment by Flakebi** [comment]
Thank you for the reviews and merging this!
